reco_utils/recommender/fbt/fbt.py

```diff
@@ -0,0 +1,311 @@
+
+    def fit(self, df):
+        """Fit the FBT recommender using an input DataFrame.
+
+        Fitting of model involves computing a item-item co-occurrence
+        matrix: how many people bought watched a pair of movies?
+
+        Args:
+            df (pd.DataFrame): DataFrame of users and items
+
+        Returns:
+        self
+            Fitted estimator.
+        """
+        logger.info("Check dataframe is of the type, schema we expect")
+        expected_cols_df = [self.col_user,
+                            self.col_item,
+                            f'{self.col_item}_name']
+        self._check_dataframe(df, expected_columns=expected_cols_df)
+
+        # copy the DataFrame to avoid modification of the input
+        select_columns = [self.col_user, self.col_item]
+        temp_df = df[select_columns].copy()
+
+        logger.info("De-duplicating the user-item counts")
+        temp_df = temp_df.drop_duplicates(
+                [self.col_user, self.col_item], keep="last"
+        )
+
+        # To compute co-occurrence, key piece is the self-join
+        cooccurrence_df = (
+            temp_df
+            .merge(temp_df, on=self.col_user,
+                   how='inner', suffixes=['', '_paired'])
+        )
+        # similarity score is defined by how many distinct
+        # users interacted with the same pair of items
+        sim_df = (
+            cooccurrence_df
+            .groupby([self.col_item,
+                     f'{self.col_item}_paired'])[self.col_user]
+            .nunique()
+            .reset_index(drop=False)
+            .rename(columns={self.col_user: self.col_score})
+        )
+
+        # Retaining item-pairs where item isn't paired with itself
+        fbt_df = (
+            sim_df
+            .loc[sim_df
+                 [self.col_item] !=
+                 sim_df[f'{self.col_item}_paired']]
+        )
+
+        self._model_df = fbt_df
+
+        # Item frequencies can be obtained by looking at the
+        # number of distinct users who interacted with a specific
+        # item: can be extracted from items paired with itself
+        self.item_frequencies = (
+            sim_df
+            .loc[sim_df
+                 [self.col_item] ==
+                 sim_df[f'{self.col_item}_paired']]
+            .drop(columns=[f'{self.col_item}_paired'])
+        )
+
+        self._is_fit = True
+        logger.info("Done training")
+
+    def predict(self, test):
+        """Generate new recommendations using a trained FBT model.
+
+        Parameters
+        ----------
+        test : DataFrame
+            DataFrame of users who need to get recommendations
+        remove_seen (bool): flag to remove items seen in training
+        from recommendation
+
+        Returns
+        -------
+        R : DataFrame
+            DataFrame with up to `num_recos` rows per user in X.
+        """
+        if not self._is_fit:
+            raise ValueError(("fit() must be called before predict()!"))
```

Comment:
typo: double parantesis

Reply:
It's not a typo :) - it's just, strictly speaking, unnecessary in this line.
Generally we wrap strings into parentheses so that if we need to split them across several lines, the strings are automatically concatenated across lines. 
Again, in this case it's not strictly needed, it's just a pattern I've seen people employ often.

Reply:
This is rarely used though when the string fits in a single line.

Reply:
resolved.